data/redis.go

```diff
@@ -80,32 +80,107 @@ func (r *RedisConnector) Id() string {
 func (r *RedisConnector) connectTask(ctx context.Context) error {
-	options := &redis.Options{
-		Addr:         r.cfg.Addr,
-		Password:     r.cfg.Password,
-		DB:           r.cfg.DB,
-		PoolSize:     r.cfg.ConnPoolSize,
-		DialTimeout:  r.initTimeout,
-		ReadTimeout:  r.getTimeout,
-		WriteTimeout: r.setTimeout,
-	}
-
-	if r.cfg.TLS != nil && r.cfg.TLS.Enabled {
-		tlsConfig, err := common.CreateTLSConfig(r.cfg.TLS)
+	var options *redis.Options
+	var err error
+	redisURI := strings.TrimSpace(r.cfg.URI)
+
+	if redisURI == "" {
+		// Construct from discrete fields (addr, username, password, db)
+		if strings.TrimSpace(r.cfg.Addr) == "" {
+			return fmt.Errorf("missing Redis connection information: either 'uri' or 'addr' must be supplied")
+		}
+
+		var userInfo string
+		if r.cfg.Username != "" || r.cfg.Password != "" {
+			userInfo = r.cfg.Username + ":" + r.cfg.Password + "@"
+		}
+
+		// Use rediss:// prefix if TLS is enabled in config, otherwise redis://
+		scheme := "redis://"
+		if r.cfg.TLS != nil && r.cfg.TLS.Enabled {
+			scheme = "rediss://"
+		}
+
+		redisURI = fmt.Sprintf("%s%s%s/%d", scheme, userInfo, r.cfg.Addr, r.cfg.DB)
+		r.logger.Debug().Str("uri", util.RedactEndpoint(redisURI)).Msg("constructed Redis URI from discrete config fields")
+	} else {
+		r.logger.Debug().Str("uri", util.RedactEndpoint(redisURI)).Msg("attempting to connect to Redis using provided URI")
+	}
+
+	options, err = redis.ParseURL(redisURI)
+	if err != nil {
+		return fmt.Errorf("failed to parse Redis URI: %w", err)
+	}
+
+	// Config overrides win only when the URI (or query params) did not already
+	// set an explicit value.
+	if r.initTimeout > 0 && options.DialTimeout == 0 {
+		options.DialTimeout = r.initTimeout
+	}
+	if r.getTimeout > 0 && options.ReadTimeout == 0 {
+		options.ReadTimeout = r.getTimeout
+	}
+	if r.setTimeout > 0 && options.WriteTimeout == 0 {
+		options.WriteTimeout = r.setTimeout
+	}
+	if r.cfg.ConnPoolSize > 0 && options.PoolSize == 0 {
+		options.PoolSize = r.cfg.ConnPoolSize
+	}
+
+	/**
+	* if tls.enabled: true in config → build a tls.Config and
+	* apply or merge it.
+	*
+	* Otherwise, keep whatever came from the URI (rediss:// or none).
+	* redis.ParseURL("rediss://…") inserts a minimal tls.Config whose only special bit is
+	* InsecureSkipVerify = true (i.e. accept any server-cert)
+	 */
+	if cfgTLS := r.cfg.TLS; cfgTLS != nil && cfgTLS.Enabled {
+		tlsConfig, err := common.CreateTLSConfig(cfgTLS)
 		if err != nil {
 			return fmt.Errorf("failed to create TLS config: %w", err)
 		}
-		options.TLSConfig = tlsConfig
+
+		if options.TLSConfig == nil {
+			// URI was redis:// — take YAML config wholesale.
+			options.TLSConfig = tlsConfig
+			r.logger.Debug().Msg("enabled TLS via YAML configuration")
+		} else {
+			// URI was rediss:// — merge YAML extras onto the baseline.
+			if len(tlsConfig.Certificates) > 0 {
+				options.TLSConfig.Certificates = tlsConfig.Certificates
+				options.TLSConfig.InsecureSkipVerify = false
```

Comment:
Should we leave that to whatever was parsed from the URI? or maybe let the configured value to override?

In situations where people use self-issued certs for internal uses

Reply:
do you mean the `InsecureSkipVerify`?

Reply:
I see, so we're assuming if certificates are provided user for sure doesn't intent to skip verification. I think it makes sense.